aten/src/ATen/native/vulkan/api/Context.cpp

```diff
@@ -0,0 +1,281 @@
+    VK_CHECK(vkCreateDebugReportCallbackEXT(
+        instance,
+        &debugReportCallbackCreateInfo,
+        nullptr,
+        &debug_report_callback));
+  }
+
+  return instance;
+}
+
+VkPhysicalDevice acquire_physical_device(const VkInstance instance) {
+  uint32_t device_count = 0;
+  VK_CHECK(vkEnumeratePhysicalDevices(instance, &device_count, nullptr));
+  TORCH_CHECK(device_count > 0, "Vulkan: Could not find a device with vulkan support!");
+
+  std::vector<VkPhysicalDevice> devices(device_count);
+  VK_CHECK(vkEnumeratePhysicalDevices(instance, &device_count, devices.data()));
+
+  return devices[0];
+}
+
+VkPhysicalDeviceLimits query_physical_device_physical_device_limits(
+    const VkPhysicalDevice physical_device) {
+  VkPhysicalDeviceProperties physical_device_properties{};
+  vkGetPhysicalDeviceProperties(physical_device, &physical_device_properties);
+  return physical_device_properties.limits;
+}
+
+uint32_t query_compute_queue_family_index(const VkPhysicalDevice physical_device) {
+  uint32_t queue_family_count = 0;
+
+  vkGetPhysicalDeviceQueueFamilyProperties(
+      physical_device, &queue_family_count, nullptr);
+
+  TORCH_CHECK(
+      queue_family_count > 0, "Vulkan: Invalid number of queue families!");
+
+  std::vector<VkQueueFamilyProperties> queue_families_properties(
+    queue_family_count);
+
+  vkGetPhysicalDeviceQueueFamilyProperties(
+      physical_device, &queue_family_count, queue_families_properties.data());
+
+  for (uint32_t i = 0; i < queue_families_properties.size(); ++i) {
+    const VkQueueFamilyProperties& properties = queue_families_properties[i];
+    if (properties.queueCount > 0 && (properties.queueFlags & VK_QUEUE_COMPUTE_BIT)) {
+      return i;
+    }
+  }
+
+  TORCH_CHECK(
+      false, "Vulkan: Could not find a queue family that supports operations!");
+}
+
+VkDevice create_device(
+    const VkPhysicalDevice physical_device,
+    const uint32_t compute_queue_family_index) {
+  const float queue_priorities = 1.0f;
+  const VkDeviceQueueCreateInfo device_queue_create_info{
+    VK_STRUCTURE_TYPE_DEVICE_QUEUE_CREATE_INFO,
+    nullptr,
+    0u,
+    compute_queue_family_index,
+    1u,
+    &queue_priorities,
+  };
+
+  const VkDeviceCreateInfo device_create_info{
+    VK_STRUCTURE_TYPE_DEVICE_CREATE_INFO,
+    nullptr,
+    0u,
+    1u,
+    &device_queue_create_info,
+    0u,
+    nullptr,
+    0u,
+    nullptr,
+  };
+
+  VkDevice device{};
+  VK_CHECK(vkCreateDevice(physical_device, &device_create_info, nullptr, &device));
+
+  return device;
+}
+
+VkQueue acquire_queue(
+    const VkDevice device,
+    const uint32_t compute_queue_family_index) {
+  VkQueue queue{};
+  vkGetDeviceQueue(device, compute_queue_family_index, 0, &queue);
+  return queue;
+}
+
+} // namespace
+
+Context::Context(const bool enable_validation_layers)
+    : instance_(create_instance(enable_validation_layers), &VK_DELETER(Instance)),
+      physical_device_(acquire_physical_device(instance())),
```

Comment:
In this inits there are some places with using member functions to get `instance()` or member field instance_.
Could it be member field everywhere?

Reply:
You mean in the constructor or in general?  I tried to only use the accessor functions in the constructor for no particular reason, but yeah, only meant to consistently use the accessors and not mix with member variables - if that's the case I missed it, could you point it out?  Functions above and in the anonymous namespace are free functions by the way and access their input and output only through their parameters so if you refer to those that's why they don't use member functions.